File: Project-Demo/SignupViewController.m
#import "SignupViewController.h"

@interface SignupViewController ()

@property (strong, nonatomic) UITextField *emailTextField;
@property (strong, nonatomic) UITextField *passwordTextField;
@property (strong, nonatomic) UIButton *signUpButton;
@property (strong, nonatomic) UIButton *backButton;

@end

@implementation SignupViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.view.backgroundColor = [UIColor whiteColor];
    
    // Set up the UI elements
    [self setupUI];
}

- (void)setupUI {
    // Back button
    UIButton *crossButton = [UIButton buttonWithType:UIButtonTypeSystem];
    crossButton.frame = CGRectMake(20, 40, 40, 40);
    [crossButton setTitle:AppConstants.BACK_ICON forState:UIControlStateNormal]; 
    [crossButton addTarget:self action:@selector(crossButtonTapped:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:crossButton];
    
    // Email Text Field
    self.emailTextField = [[UITextField alloc] initWithFrame:CGRectMake(40, 150, self.view.frame.size.width - 80, 40)];
    self.emailTextField.placeholder = Messages.ENTER_EMAIL;
    self.emailTextField.borderStyle = UITextBorderStyleRoundedRect;
    self.emailTextField.autocapitalizationType = UITextAutocapitalizationTypeNone;
    [self.view addSubview:self.emailTextField];
    
    // Password Text Field
    self.passwordTextField = [[UITextField alloc] initWithFrame:CGRectMake(40, 200, self.view.frame.size.width - 80, 40)];
    self.passwordTextField.placeholder = Messages.ENTER_PASSWORD;
    self.passwordTextField.borderStyle = UITextBorderStyleRoundedRect;
    self.passwordTextField.secureTextEntry = YES;
    [self.view addSubview:self.passwordTextField];
    
    // Login Button
    self.signUpButton = [UIButton buttonWithType:UIButtonTypeSystem];
    self.signUpButton.frame = CGRectMake(40, 260, self.view.frame.size.width - 80, 40);
    [self.signUpButton setTitle:Messages.SIGNUP forState:UIControlStateNormal];
    [self.signUpButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    self.signUpButton.backgroundColor = [UIColor systemBlueColor];
    self.signUpButton.layer.cornerRadius = 5;
    [self.signUpButton addTarget:self action:@selector(signUpButtonTapped:) forControlEvents:UIControlEventTouchUpInside];
    [self.view addSubview:self.signUpButton];
}

// Action for Login Button
- (void)signUpButtonTapped:(id)sender {
    NSString *email = self.emailTextField.text;
    NSString *password = self.passwordTextField.text;
    
    if (email.length == 0 || password.length == 0) {
        [self showAlertWithTitle:Messages.ERROR message:Messages.ENTER_EMAIL_AND_PASSWORD];
    } else {
        [FirebaseAuthManager.shared signUpWithEmail:email password:password completion:^(FIRAuthDataResult * _Nullable authResult, NSError * _Nullable error) {
            if(error){
                [self showAlertWithTitle:Messages.ERROR message:error.localizedDescription];
            } else {
                [self presentHomeViewController];
            }
        }];
    }
}

- (void)presentHomeViewController {
    UIStoryboard *storyboard = [UIStoryboard storyboardWithName:AppConstants.STORY_BOARD_NAME bundle:nil];
    HomeViewController *homeVC = [storyboard instantiateViewControllerWithIdentifier:AppConstants.HOME_NAVIGATION_VIEW_CONTROLLER];
    homeVC.modalPresentationStyle = UIModalPresentationFullScreen;
    [self presentViewController:homeVC animated:YES completion:nil];
}


- (void)crossButtonTapped:(id)sender {
    [self dismissViewControllerAnimated:YES completion:nil];
}

- (void)showAlertWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:title
                                                                             message:message
                                                                      preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction *okAction = [UIAlertAction actionWithTitle:Messages.OK
                                                       style:UIAlertActionStyleDefault
                                                     handler:nil];
    [alertController addAction:okAction];
    [self presentViewController:alertController animated:YES completion:nil];
}

@end

File: SignUpViewController.swift
import UIKit

protocol SignupViewControllerProtocol : AnyObject {
    func invalidInput() -> Void
    func failedWithError(_ error : Error) -> Void
    func signUpSucceded() -> Void
}

class SignupViewController: UIViewController,SignupViewControllerProtocol {

    private var emailTextField: UITextField!
    private var passwordTextField: UITextField!
    private var signUpButton: UIButton!
    
    var viewModel: SignupViewModel?
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.view.backgroundColor = .white
        
        // Set up UI elements
        setupUI()
    }

    private func setupUI() {
        // Back button
        let crossButton = UIButton(type: .system)
        crossButton.frame = CGRect(x: 20, y: 40, width: 40, height: 40)
        crossButton.setTitle(AppConstants.BACK_ICON, for: .normal)
        crossButton.addTarget(self, action: #selector(crossButtonTapped(_:)), for: .touchUpInside)
        self.view.addSubview(crossButton)
        
        // Email Text Field
        emailTextField = UITextField(frame: CGRect(x: 40, y: 150, width: self.view.frame.size.width - 80, height: 40))
        emailTextField.placeholder = Messages.ENTER_EMAIL
        emailTextField.borderStyle = .roundedRect
        emailTextField.autocapitalizationType = .none
        self.view.addSubview(emailTextField)
        
        // Password Text Field
        passwordTextField = UITextField(frame: CGRect(x: 40, y: 200, width: self.view.frame.size.width - 80, height: 40))
        passwordTextField.placeholder = Messages.ENTER_PASSWORD
        passwordTextField.borderStyle = .roundedRect
        passwordTextField.isSecureTextEntry = true
        self.view.addSubview(passwordTextField)
        
        // Sign Up Button
        signUpButton = UIButton(type: .system)
        signUpButton.frame = CGRect(x: 40, y: 260, width: self.view.frame.size.width - 80, height: 40)
        signUpButton.setTitle(Messages.SIGNUP, for: .normal)
        signUpButton.setTitleColor(.white, for: .normal)
        signUpButton.backgroundColor = .systemBlue
        signUpButton.layer.cornerRadius = 5
        signUpButton.addTarget(self, action: #selector(signUpButtonTapped(_:)), for: .touchUpInside)
        self.view.addSubview(signUpButton)
    }
    
    // Action for Sign Up Button
    @objc private func signUpButtonTapped(_ sender: UIButton) {
        self.viewModel?.signup(email: emailTextField.text ?? "", password:passwordTextField.text ?? "")
    }
    
    // Present HomeViewController
    private func presentHomeViewController() {
        let storyboard = UIStoryboard(name: AppConstants.STORY_BOARD_NAME, bundle: nil)
        let homeVC = storyboard.instantiateViewController(withIdentifier: AppConstants.HOME_NAVIGATION_VIEW_CONTROLLER)
        homeVC.modalPresentationStyle = .fullScreen
        self.present(homeVC, animated: true, completion: nil)
    }
    
    // Action for Back Button
    @objc private func crossButtonTapped(_ sender: UIButton) {
        self.dismiss(animated: true, completion: nil)
    }
    
    // Show Alert
    private func showAlert(title: String, message: String) {
        let alertController = UIAlertController(title: title, message: message, preferredStyle: .alert)
        let okAction = UIAlertAction(title: Messages.OK, style: .default, handler: nil)
        alertController.addAction(okAction)
        self.present(alertController, animated: true, completion: nil)
    }
    
    func invalidInput() {
        showAlert(title: Messages.ERROR, message: Messages.ENTER_EMAIL_AND_PASSWORD)
    }
    
    func signUpSucceded() {
        presentHomeViewController()
    }
    
    func failedWithError(_ error: Error) {
        showAlert(title: Messages.ERROR, message: error.localizedDescription)
    }
}

